### PR TITLE
reflex_name restriction loosening

### PR DESCRIPTION
**Pull request by leastbad**
## Type of PR (feature, enhancement, bug fix, etc.)

# Enhancement

## Description

This PR would make the string fragment "Reflex" used in the `data-reflex` attributes optional, allowing for a shorter syntax.

It also allows for the use of Reflex classes that do not end in Reflex, while not changing the existing default behavior.

### Examples

This is the current behavior:

```html
<button data-reflex="click->PowerReflex#lift">Lift</button>
```
This calls PowerReflex#lift.

```html
<button data-reflex="click->Power#lift">Lift</button>
```
This first checks to see if there is a Power class, and if there is, that it implements `StimulusReflex::Reflex`. If everything looks good, Power is your Reflex class.

If there is no Power class, or Power does not inherit from `StimulusReflex::Reflex`, it falls back to PowerReflex as your class. This is transparent to the client.

## Why should this be added

First, I should say that I'm not emotionally attached to this outcome, and I'm not in love with my implementation - although I'm glad it's 100% server-side. Unfortunately, Ruby's `defined?` method handles `User` and `"User".constantize` differently:

```ruby
x = User
defined? User # true
defined? "User".constantize # "method"
defined? x # "local-variable"
```
I'm sure there are more experienced Rubyists who know how to make this better, but I didn't want to require therapy to get this done, so I went with an admittedly hacky begin..rescue approach, instead.

Anyhow, here are the FOR arguments:

- shorter syntax
- not restricted to classes ending in "Reflex"
- backwards compatible

## Heads up

I found a bug: line 31 of `channel.rb` (on current master branch; it's 32 in this PR) raises an ArgumentError if the reflex_class does not inherit from `StimulusReflex::Reflex`. However, line 35/36 - the first line of the error rescue, calls `reflex.rescue_with_handler()`. Unfortunately, if reflex_class is *not* a Reflex, it will be **nil**, causing this rescue mission to fail.

In other words, on today's master branch, if we have a power_reflex.rb that forgets to inherit, it will throw this ugly error instead of a nice friendly error.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] Checks (StandardRB & Prettier-Standard) are passing

### CODE REVIEW

File: lib/stimulus_reflex/channel.rb [hunk 1]
```diff
@@ -52,6 +53,13 @@ def receive(data)
 
   private
 
+  def compute_reflex_name(reflex_name)
+    raise unless is_reflex?(reflex_name.constantize)
+    reflex_name
+  rescue
+    "#{reflex_name}Reflex"
```
Comment (foca) [inline lib/stimulus_reflex/channel.rb:60]: You can use `const_defined?` for this:

```suggestion
    Object.const_defined?(reflex_name.to_s) ? reflex_name : "#{reflex_name}Reflex"
```

Needs to be scoped to `Object` to ensure the lookup path is what you'd expect. And _needs_ to be a String, hence the `to_s` (technically, anything that responds `to_str`.)

File: lib/stimulus_reflex/channel.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ def receive(data)
     target = data["target"].to_s
     reflex_name, method_name = target.split("#")
     reflex_name = reflex_name.classify
+    reflex_name = reflex_name.end_with?("Reflex") ? reflex_name : compute_reflex_name(reflex_name)
```
Comment (hopsoft) [inline lib/stimulus_reflex/channel.rb:25]: What about something like this?

```ruby
reflex_name = reflex_name.end_with?("Reflex") ? reflex_name : "#{reflex_name}Reflex"

# ...

reflex_class = reflex_name.constantize.tap { |klass| raise unless is_reflex?(klass) }
```

File: lib/stimulus_reflex/channel.rb (hunk 1 above)
Comment (leastbad) [reply lib/stimulus_reflex/channel.rb:60]: There needs to be a #clearlydoesntknowtherightfunction tag. You are very kind. Obviously you are right!

File: lib/stimulus_reflex/channel.rb (hunk 2 above)
Comment (leastbad) [reply lib/stimulus_reflex/channel.rb:25]: This would work, but it has less functionality than what I proposed.

My version lets you define a class with any name that inherits from `StimulusReflex::Reflex`

So you could have a totally viable `Balloon << StimulusReflex::Reflex`

Comment (hopsoft) [reply lib/stimulus_reflex/channel.rb:25]: Not sure I like that flexibility. Imagine a Rails controller not suffixed with “Controller”. I’d rather enforce convention on this.

Comment (leastbad) [reply lib/stimulus_reflex/channel.rb:25]: While I rarely find myself accessing controller consts directly, I agree with your point overall and then some. If the tables were turned, I probably wouldn't accept this logic. I do like putting sketches on the table in case something sticks. :)

**Comment by leastbad** [comment]
Note that I'm pretty sure that this still explodes the rescue attempt if reflex_class is somehow not a valid SR:R, but perhaps this is as fine as it's always been. @hopsoft

File: lib/stimulus_reflex/channel.rb (hunk 1 above)
Comment (leastbad) [reply lib/stimulus_reflex/channel.rb]: @foca we went in a different, simpler direction with this but I really appreciate learning about const_defined? !

Comment (foca) [reply lib/stimulus_reflex/channel.rb]: The resulting code is much simpler overall! I digs.
